net/dns/manager_windows.go

```diff
@@ -111,6 +115,55 @@ func (m windowsManager) setSplitDNS(resolvers []netip.Addr, domains []dnsname.FQ
 	return m.nrptDB.WriteSplitDNSConfig(servers, domains)
 }
 
+// setHosts sets the hosts file to contain the given host entries.
+func (m windowsManager) setHosts(hosts []*HostEntry) error {
+	b, err := os.ReadFile(`C:\Windows\System32\drivers\etc\hosts`)
+	if err != nil {
+		return err
+	}
+	sc := bufio.NewScanner(bytes.NewReader(b))
+	const (
+		header  = "# TailscaleHostsSectionStart"
+		comment = "# This section contains MagicDNS entries for Tailscale.\n" +
+			"# Do not edit this section manually.\n"
+		footer = "# TailscaleHostsSectionEnd"
+	)
+	var out bytes.Buffer
+	var inSection bool
+	for sc.Scan() {
+		line := strings.TrimSpace(sc.Text())
+		if line == header {
+			if !inSection {
+				inSection = true
+			}
+			continue
+		}
+		if line == footer {
+			if inSection {
+				inSection = false
+			}
+			continue
+		}
+		if inSection {
+			continue
+		}
+		fmt.Fprintln(&out, line)
+	}
+	if err := sc.Err(); err != nil {
+		return err
+	}
+	if len(hosts) > 0 {
+		fmt.Fprintln(&out, header)
+		fmt.Fprintln(&out)
+		for _, he := range hosts {
+			fmt.Fprintf(&out, "%s\t%s\n", he.Addr, strings.Join(he.Hosts, "\t"))
+		}
+		fmt.Fprintln(&out)
+		fmt.Fprintln(&out, footer)
+	}
+	return atomicfile.WriteFile(`C:\Windows\System32\drivers\etc\hosts`, out.Bytes(), 0 /* ignored */)
```

Comment:
Is this going to cause problems if we don't use the Windows `\r\n` line ending convention? At the very least, it'll make it annoying if someone wants to edit this file with Notepad or something, I would think.

Reply:
Good catch, done.